grayscale.py

```diff
@@ -392,97 +398,88 @@ def brightness_sync(self, c):
-                if state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Stopping:
-                    self._teardown_clear_render_buffer1()
-                    self.dc(1)
-                    self.spi.write(self._buffer1)
-                    state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Stopped
-                    return
-                if state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Pausing:
-                    state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Paused
-                    while state[Grayscale_StateIndex_State] != Grayscale_ThreadState_Resuming:
-                        idle()
-                    state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Running
-        except Exception as e:
-            self._state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Exception
-            print(e)
+        b1:ptr32 = ptr32(self.buffer1) ; b2:ptr32 = ptr32(self.buffer2)
+        _b1:ptr32 = ptr32(self._buffer1) ; _b2:ptr32 = ptr32(self._buffer2) ; _b3:ptr32 = ptr32(self._buffer3)
 
+        fn:int ; i:int ; t0:int
+        v1:int ; v2:int ; contrast:int
 
-    @micropython.viper
-    def _teardown_clear_render_buffer1(self):
-        b1:ptr32 = ptr32(self._buffer1)
-        i:int = 0
-        while i < 90:
-            b1[i] = 0
-            i += 1
+        state[Grayscale_StateIndex_State] = Grayscale_ThreadState_Running
+        while True:
+            while state[Grayscale_StateIndex_State] == Grayscale_ThreadState_Running:
+                fn = 0
+                while fn < 3:
+                    t0 = ticks_us()
+                    dc(0)
+                    spi_write(pre_frame_cmds)
+                    dc(1)
+                    spi_write(buffers[fn])
+                    dc(0)
+                    spi_write(post_frame_adj[fn])
+                    sleep_us(Grayscale_pre_frame_time_us - int(ticks_diff(ticks_us(), t0)))
+                    t0 = ticks_us()
+                    spi_write(post_frame_cmds)
+                    spi_write(post_frame_adj[fn])
+                    if (fn == 2) and (state[Grayscale_StateIndex_CopyBuffs] != 0):
+                        i = 0
+                        while i < 90:
+                            v1 = b1[i]
+                            v2 = b2[i]
+                            _b1[i] = v1 | v2
+                            _b2[i] = v2
+                            _b3[i] = v1 & v2
+                            i += 1
+                        state[Grayscale_StateIndex_CopyBuffs] = 0
+                    elif (fn == 2) and (state[Grayscale_StateIndex_ContrastChng] != 0xffff):
+                        contrast = state[Grayscale_StateIndex_ContrastChng]
+                        state[Grayscale_StateIndex_ContrastChng] = 0xffff
+                        post_frame_adj[0][1] = contrast >> 5
+                        post_frame_adj[1][1] = contrast >> 1
+                        post_frame_adj[2][1] = (contrast << 1) + 1
```

Comment:
Could be my dumb brain not understanding 🙈 But it seems like you're using different operations on the contrast here than in the initial setting at line 208?

Reply:
Yeah, I think this is just a holdover from going from the ssd1306.py-like `contrast()` and moving to the thumby.py-like `brightness()`. As you noted elsewhere, the latter uses a max of 127, whereas the former (and the controller use a max of 255).
The reason for the discrepancy here is that the code you've highlighted is for a range of 0 to 127, but the init code is:
```
        # but with a set of contrast values spanning the entire range provided by the controller
        brightnessVals = [0,112,255]
        brightnessVal = brightnessVals[brightnessSetting]
        # 0x81,<val>        Set Bank0 contrast value to <val>
        self.post_frame_adj = [bytearray([0x81,brightnessVal>>6]), bytearray([0x81,brightnessVal>>1]), bytearray([0x81,brightnessVal])]
```
and you can see the different shifts are used but with a set of brightness/contrast values that are in the range 0 to 255.

I can change the init code to match to avoid confusion.